### PR TITLE
feat: Multi Language Surveys

### PR DESCRIPTION
**Pull request by Dhruwang**
## What does this PR do?

Allow user to create and answer surveys in multiple languages

Fixes [855](https://linear.app/formbricks/issue/FOR-855/status-multi-language-support)
Screenshot 2023-11-17 at 12 08 39 PM
Screenshot 2023-11-17 at 12 09 19 PM

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Chore (refactoring code, technical debt, workflow improvements)
- [ ] Enhancement (small improvements)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change adds a new database migration
- [ ] This change requires a documentation update

## How should this be tested?

Go to survey editor and click on Add translation
Toggle langauge buttons

You can change your langauge settings in settings/language

NOTE Its an enterprise feature so you need to set your `ENTERPRISE_LICENSE_KEY` in order to use it

### Required

- [x] Filled out the "How to test" section in this PR
- [x] Read [How we Code at Formbricks](<[https://github.com/formbricks/formbricks/blob/main/CONTRIBUTING.md](https://formbricks.com/docs/contributing/how-we-code)>)
- [x] Self-reviewed my own code
- [x] Commented on my code in hard-to-understand bits
- [x] Ran `pnpm build`
- [x] Checked for warnings, there are none
- [x] Removed all `console.logs`
- [x] Merged the latest changes from main onto my branch with `git pull origin main`
- [x] My changes don't cause any responsiveness issues
- [x] First PR at Formbricks? [Please sign the CLA!](https://formbricks.com/clmyhzfrymr4ko00hycsg1tvx) Without it we wont be able to merge it 🙏

### Appreciated

- [x] If a UI change was made: Added a screen recording or screenshots to this PR
- [x] Updated the Formbricks Docs if changes were necessary

### CODE REVIEW

**Comment by Dhruwang** [comment]
@jobenjada Have updated the PR 😊
Also for taking a survey in a specific language , we can just append ?lang=LANGAUGE_SYMBOL with the link

**Comment by jobenjada** [comment]
Hey @Dhruwang 

going over it again, I noticed the following:

1. Add red to show which fields arent validated:
image

Have a chat with @ShubhamPalriwala - he can help you get it done quickly since he built the validation :)

Ideally, we would also change the toast message to: "Translations missing, please complete them to save."

-------

2. Move Language Switch into Questions View

 There is no need to have it in the Settings Tab as well. It suggests that the settings are different for each language which is not true

image

--------

3. Do these props make sense? 🤓 

Seems  not clearly distinguishable to me:

image

------

4. UX of adding language isn't great mostly because once you added it in the new tab, its not added in the Survey Editor without a hard reload. Here is an idea:

- When the "Add language" button is clicked, and the Survey Editor tab is selected again, we assume that the user has added a new language so we save the current status and perform the reload automatically.

Maybe @mattinannt has a better idea :D

--------

6. Lastly, the "Add Translation" button position. I would like to do the following:

On SurveyEditor load, the scroll depth is here, so the Add Translation button is hidden:

image

And when people work with the editor and scroll up and down, they will find the button:

image

This way we keep the UI clean and easy for new users and  still have the button readily available when people need it :)

------

Thats it in terms of UI/UX - please resolve the merge conflicts so I can test the functionality through. Please do the same after fixing merge conflcits so we can get all the bugs before this goes into production 🙏

**Comment by jobenjada** [comment]
hey @Dhruwang - looks good :) Great work with the live sync when a new question is added 💪 

I have noticed a few more things:

1. Placeholder not multi language yet:
image

-----------

2. Rich Text not multi language yet (see Slack thread):

image

------

2b. Button Texts not multi language yet
image


----------

3. What about the last point I made re scroll depth? Is that too hacky? Let's discuss on Slack.

--------

4. When I switch tabs, the button goes back to default even though there is a translation added:
image

---------

Haven't tested funtionality again since we might switch up the editor. Will do it after that :)

Great work! 🚀

**Comment by jobenjada** [comment]
Hey @Dhruwang 

here is my feedback:

1. Width if off - there is horizontal scroll:

image

------------

2. Rich Text Area doesn't update survey text (known bug):

image

------------

3. Welcome Card Headline misses translation - **please check all input fields of all questions!**

image

-----------

4. same for the Next button on Welcome Card

image

--------

5. "Other" option breaks app

image
image

---------

6. Placeholder field misses translations

image


@Dhruwang See Slack message

**Comment by jobenjada** [comment]
Hey @Dhruwang - here is how we'll tackle the multi language X recall functionality issues. Here are solutions to the edge cases you raised.

**Missing Recall in one language**
Suppose a user adds question in english and adds a recall in it. Now he enables translation (lets say german) and write same headline in german but misses to add recall

Correct me if I'm wrong, but I think we can leave this to the user. If they want to add a Recall in all languages, they can, but if they don't want it's also ok. I don't think it would break the form or effect the analysis, because it only effects the survey questions and description.

The only case we should design for is users forgetting to add the Recall in the translation. So from this translate hint it should become clear that they should add the Recall also to this question:

image

---------

**Fallback language**
For a multi language survey with recall question, should the fallback be different for different languages?

Yes, we should allow users to add the fallback in several language using the same language switcher other free text inputs have:

image

**Validation**
Please make sure to add senseful validation rules :)

-----------

Thanks!! 🚀

**Comment by jobenjada** [comment]
Hey Dhru!

Looks really good, wasn't able to break it 🤓  great work! Here is some UX feedback:

1. Share Modal: Add language select to LinkTab.tsx

I've tweaked the UI a bit, pls add the same to the LinkTab.tsx

image

------

2. Clean ups in LinkTab.tsx

------

3. I removed the languages and the dropdown refreshed but the survey editor did not:

image

I think in this case we could justify a hard reload, if it cannot be avoided

------

4. Some errors in EditLanguage.tsx

image

-----

5. Sometimes there seems to be an issue with changes not being saved properly when the validation blocks it. Not entirely sure when it happens tho.

https://github.com/formbricks/formbricks/assets/72809645/1527d7a7-5f7a-4ce5-98dd-5b72d17d2af5


-----

6. Validation Messages not specific enough

I'd love you to spend a bit more time on trying to write more specific error messages for the validation. It quickly happens that the whole survey editor lights up in red and users don't really understand why. Not entirely sure how the validation logic works so please see what you can do 🤓  Look for the pareto solution: 80% better validation messages with 20% of the work :) 

------

Apart from this, it looks super good!!!

Thanks!

**Comment by mattinannt** [comment]
@Dhruwang I started fixing the build errors but there seems to be a few more. Please fix all build errors and build warnings first, then we can continue with the review 😊

File: apps/web/app/(app)/environments/[environmentId]/surveys/[surveyId]/edit/components/RatingQuestionForm.tsx
Comment (mattinannt) [inline apps/web/app/(app)/environments/[environmentId]/surveys/[surveyId]/edit/components/RatingQuestionForm.tsx:43]: Also here we need to be cautious with the type conversions. As this is ignoring typescript. We should a better way.

File: packages/types/tsconfig.json
Comment (mattinannt) [inline packages/types/tsconfig.json:3]: I'm not sure if we are able from a licensing point of view to import code from `ee` into the surveys or js package. We shouldn't do this if we can avoid this. If they are general types and general lib functions please move them to the `types` and `lib` packages.

**Review by mattinannt — CHANGES_REQUESTED** [review]
@Dhruwang The feature looks pretty great overall 😊🔥 Great job 👏
It's a really complex but powerful feature that also changes some of the fundamentals how Formbricks is currently working (e.g. the string types).

I currently see a few issues left with the current state:
- You often used `"en"` as a default language identifier but this might not work since the user can change the identifier
- You often do type conversions to `TI18nString`, but from the rest of the typescript types this seems inaccurate and ignores the typescript checking that should help us and could lead to errors down the road.
- We need to make sure we can provide backwards compatibility with older versions of `@formbricks/js`
- We need to add license checks for the Cloud licenses (active inApp subscription)
- We need to add tests
- We are unable to import types or functions from the `ee` to the surveys or js package due to license issues. Please move all general functions and types that are necessary for using Formbrocks  to the `lib` and `types` package instead :-) 

I also thought about transforming all fields in the existing surveys to i18n fields as this might make the types easier. Please reach out to @pandeymangg as he is currently building our first data migration for the advanced user targeting feature. If we build a data migration for the multi language feature, too, this could be great (now that we know how to do this).

For tests we need to check again what makes sense. I think a few unit tests for the utils is something we should definitely do to make sure this functionality is working properly.
If we can also create one e2e test that tests a simple multi-language survey in creation, that would be great :-)

**Comment by Nick-v-L** [comment]
Hi @Dhruwang,
I was just checking this PR out, because I really like the features. Great work! Was checking it out on my local machine and found the following error (see video).
https://www.loom.com/share/89417d7146684eecb7dc7f0ac576e764

**Comment by Nick-v-L** [comment]
@Dhruwang, I checked it again, but unfortunately still found a few of the same issues.
https://www.loom.com/share/3b564f2e7d1c4665906096ed804bb437

**Comment by jobenjada** [comment]
## Testing Log

For my surveys, the migration worked well (Churn Survey and PMF survey templates) ✅

I have updated the UI / UX and am happy with it ✅

The only issue: I couldn't get the SDK to spit out the right survey ❌

The attribute is set (actually twice):

image

But im seeing the English survey even though the language attribute is stored correctly within Formbricks:
image

What am I doing wrong?

@Dhruwang as discussed, please have a look if it's possible to make changes to several languages at once. If not, please update my refactored version of the settings page.

Thanks 🙌

**Comment by jobenjada** [comment]
Hey Dhru! Here is my test log:

1. Issue when deleting last question, all questions reappear
https://github.com/formbricks/formbricks/assets/72809645/2951f35a-b783-47c8-b28f-a45e3f7c120c

----------

2. Survey closed message not yet translateable
image

---------

3. Verify Email not yet translateable
image

--------

4. Language Dropdown Width
image

---------

5. Single Select Options Spacing is off (too far from each other):
image

--------

6.  Schedule Question seems broken

image

it also blocks saving even though I filled out all the fields:

image

--------

7. Placeholder field: Translate preview is off when the field is empty
image
image

---------

8. Hover state is a bit off sometimes:
image

-------

9. Analysis does not consider all options (3 responses vs 1 response in the select field):
image

I am not entirely sure how to handle this @mattinannt 

--------

10. The filter doesn't seem to work on Summary page:
image

-------

11. Got this error after switching on userIdentification
image

-------

12. Not sure if the language selection works. In the top right corner you see I'm setting "en" but its showing the German survey:
image

-------

13. After I released the survey under one action (New Session) I cannot rerelease it under a different action. In the survey editor I got the "Error saving changes" Toast and this:
image

-------

14. I was about to say that the language targeting does not seem to work, but now it actually showed the survey in the right language.

It seems to be not very reliable though. I'm sure, the different caching things make it challenging, but it's not easy to test so it's hard to believe someone builds up enough trust to use it in production. So it has to be easy and clear why we see which language, ideally via the debug logs.

Here is one example: It first shows the survey in ES (like it should) and then I reopen it and then its in default language:


https://github.com/formbricks/formbricks/assets/72809645/5eae0c44-62da-4b4d-9a4e-dc8acaa0fd43


-------

15. 'undefined' log
image


-------

Thanks!

**Comment by jobenjada** [comment]
Hey Dru! Starting to look really good 😎 As discussed, these three things I noticed:

1. Don't display default if lang flag doesnt match for in app
2. Invalidate filter cache correctly
3. Resolve merge conflict

@mattinannt after this, it's done from my side! 🙌

**Review by mattinannt — APPROVED** [review]

**Comment by arghaffari** [comment]
@jobenjada If customers could change the language on demand by themselves, it would be great. Since we actually don't know which customers in different markets we are going to send surveys to, they would be able to choose their desired language with this feature.
https://surveys.feedbackly.com/survey/1r5P9ADl

**Pull request merged**
